types/node/index.d.ts

```diff
@@ -3986,6 +4046,13 @@ declare module "fs" {
      */
     export function readdir(path: PathLike, callback: (err: NodeJS.ErrnoException, files: string[]) => void): void;
 
+    /**
+     * Asynchronous readdir(3) - read a directory.
+     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
+     * @param options If called with `withFileTypes: true` the result data will be an array of Dirent.
+     */
+    export function readdir(path: PathLike, options: { withFileTypes: true }, callback: (err: NodeJS.ErrnoException, files: Dirent[]) => void): void;
```

Comment:
The `withFileTypes` option has higher prio then `encoding`. Not sure if the overload resolution picks the right one if both options are set. A test would be good.

Reply:
I did add a test, it works fine and errors if you try to specify `encoding`.

Reply:
That said, there is no test in the `$ExpectError` and I'm hesitant to use it given it seems to be broken in unrelated packages.

Reply:
what will not work is to set `withFileTypes: false` or combine `withFileTypes: false, encoding: "buffer"`. But I think these cases are of no use anyway and if really needed they can be added in a separate PR.

Reply:
Ah yea, that makes sense, lemme look into that.

Reply:
I added 1 commit and test that sort of deals with the issue but I'm not exactly happy with it, comments welcome.

Reply:
Writing type definitions for such functions results in a lot text which is for sure not nice and quite error prone but I have no better idea.